pinot-common/src/main/java/org/apache/pinot/sql/parsers/CalciteSqlParser.java

```diff
@@ -147,6 +147,9 @@ static void validate(PinotQuery pinotQuery)
   private static void validateGroupByClause(PinotQuery pinotQuery)
       throws SqlCompilationException {
     if (pinotQuery.getGroupByList() == null) {
+      if (containsAggregateAndColsInSelectClause(pinotQuery)) {
+        throw new SqlCompilationException("Columns and Aggregate functions can't co-exist without GROUP BY clause");
+      }
       return;
     }
     // Sanity check group by query: All non-aggregate expression in selection list should be also included in group
```

Comment:
Isn't this code block checking what this PR is trying to accomplish ? From the description, lines 157 to 162 check if the select expression is not an aggregate expression and also not a group by expression then raise error. So all queries like following should be caught here imo. 

If these queries are not being caught in this code block, then I think we should fix this code instead of introducing another function to do the same check

```
SELECT col1, col2, count(*) FROM foo GROUP BY col1 -- col2 is not present in group by
SELECt col1, count(*) FROM foo -- col1 not present in group by
```

Reply:
> Isn't this code block checking what this PR is trying to accomplish ? From the description, lines 157 to 162 check if the select expression is not an aggregate expression and also not a group by expression then raise error. So all queries like following should be caught here imo.
> 
> If these queries are not being caught in this code block, then I think we should fix this code instead of introducing another function to do the same check
> 
> ```
> SELECT col1, col2, count(*) FROM foo GROUP BY col1 -- col2 is not present in group by
> SELECt col1, count(*) FROM foo -- col1 not present in group by
> ```
 Correct, so logic in lines 157 to 162 assumes that there is at-least 1 group by expression and checks whether a non-aggregate expression is part of the group by clause, if not it throws an exception. But this pr handles the  case where there is no group by expression(i.e group-by expression list is null). So  I thought it would be cleaner to handle this case in a different function

Reply:
I think the fix is simpler. Making a simple change in `expressionOutsideGroupByList` function to check if `groupByExprs` is null or empty should solve help catch the query you are referring to.

Reply:
I think only adding a null or empty check to expressionOutsideGroupByList will not suffice because it will block every query which has a columnName, and the group by list is empty for example the following type of query will be blocked
```
select col1 from foo
```
because it will satisfy the blocking condition group by list empty  and not an Aggregate Expression.  for this logic to work we need to have the extra information about whether there is altleast one aggregate expression in select clause which will require an additional pass over the select list.

We can always follow that logic if we have a consensus on that would be the right way to do it.

Reply:
If possible, I would like to avoid additional pass over `pinotQuery.getSelectList()`. How about we change code in lines 157 to 163 in the following way to achieve this ?

```
boolean groupByListEmpty = pinotQuery.getGroupByList().isEmpty() || pinotQuery.getGroupByList() == null;

for (Expression selectExpression : pinotQuery.getSelectList()) {
     if (isAggregateExpression(selectExpression)) {
             aggregateExprCount++;
     }
     else if (!groupByListEmpty ) {
             if (expressionOutsideGroupByList(selectExpression, groupByExprs)) {
                  throw new SqlCompilationException(
                  "'" + RequestUtils.prettyPrint(selectExpression) + "' should appear in GROUP BY clause.");
             }
     } else if (expressionContainsColumn(expr) && aggregateExprCount  > 0) {
            throw new SqlCompilationException(
                           "Non-aggregate and Aggregate expressions can't co-exist without GROUP BY clause");
     }
}
```

Let's also remove code between lines 149 and 152

Also I feel that we don't need to explicitly check `expressionContainsColumn(expr)`. If the expression in the `selectList` is not an aggregate expression, then it could either be a non-aggregate literal/identifier/function expression. In any case, if the `aggregateExprCount  > 0`, we know there is a combination of non-aggregate and aggregate expression together in the selectList without GROUP BY clause. 

So the last else if condition can simply be `else if (aggregateExprCount  > 0)`

Reply:
Cool makes sense would make the changes on the above lines. The only thing being  we might still need call to expressionContainsColumn(expr) to allow the queries like below:
```
select 'foo' as foo, sum(c1) from bar
select upper('foo'), count(*) from bar
```
the Non-Aggregate Expression in the above doesn't have a column name  hence they must allowed(currently, this type of query are allowed and valid). let me know if that okay?

Reply:
I would suggest keeping it simple for now as outlined in my previous comment

Reply:
will make the changes